accepted/00000-http-api.md

```diff
@@ -0,0 +1,219 @@
+
+
+## 4. Exception Handling
+
+In XMLRPC, all the exceptions used inherit `FaultException` directly. Every exception has a unique integer code that describes an XMLRPC error. For HTTP, the `FaultException` tree must be enhanched with an HTTP status code that will be used in the response in any case of exception. Unlike the unique XMLRPC fault codes, HTTP codes will be chosen from a small subset of the standard HTTP status codes.
+
+Apart from the ehancement above, the current exception structure can be used as-is with the HTTP interface.
+
+
+## 5. Security
+
+The HTTP interface reuses the existing Spark framework. Therefore, we're typically not exposed to any new security issues. However, with the introduction of the new API, existing interface will be more critical for external use, so it would be beneficial to propose an approach to security testing.
+
+
+## 6. Documentation
+
+APIDOC is an extensive tool that automatically generates documentation for the API endpoints written in many languages. It can be used to generate the documentation for the HTTP API as well.
+More research is required to provide the framework for creating documentation for both interfaces with minimum duplication.
+
+
+# Unresolved Questions
+
+### Naming standard
+
+We use the term REST causally when describing the HTTP API, but as covered in this document, we don't intend to implement "REST" in a traditional way. Therefore, the term REST might be misleading for the user. On the other hand, "HTTP API" does not reflect the whole picture as well, since the existing XMLRPC API technically also utilizes HTTP. For the most clarity, we must determine a standard name for the API and use it consistently on all user-facing media (web UI, documentation, etc.).
+
+### Automatic handler discovery
+
+Currently, all the handler classes must be registered manually in the `HandlerFactory` class. This can be improved by the use of class annotations and package scanning. However, this would require use of a 3rd-party library since automatic scanning for annotations in packages is not possible with plain Java. Since the HTTP API implementation will not introduce any new handlers initially, this improvement will only provide convenience for the future additions in expense of additional work and an extra dependency in the initial implementation.
+
+### Parameter name matching
+
+XMLRPC relies strictly on the order of the sent arguments to match a method signature to a request. In contrast, the arguments sent via HTTP or JSON does not have any explicit ordering. For that reason, we must rely on parameter names and types to be able to match a signature to a request.
```

Comment:
I'm confused this. Why not passing parameter in a POST request? This will allow parameter names to be used. Take a look at this document: https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/

Reply:
The restriction is on the Java side, not HTTP.

We try to avoid using the term REST, because in nature, this API is more of an RPC over JSON than REST.

The RPC methods in question are part of the existing XMLRPC API, and they are implemented as plain Java methods in the backend. Consider this example:
```java
public String myMethod(String foo, String bar)
```
with this design, we must match the following request to this method:
```
https://uyuni.server/rhn/manager/api/myMethod?foo=value1&bar=value2
```
or this:
```
POST https://uyuni.server/rhn/manager/api/myMethod
{
  'foo': 'value1',
  'bar': 'value2'
}
```

In either form, we have the parameter names and values on the request, and the order of the parameters is arbitrary. But by default, the Java compiler does not preserve parameter names in method signatures, so in compiled code, the method becomes this:
```java
public String myMethod(String arg0, String arg1)
```
in this case, it's not possible to know which of these parameters is `foo` and which is `bar`.